File: scala/dao/implementations/MealImpl.scala
package dao.implementations

import dao.interfaces.MealInterface
import entities.Meal
import connection.DatabaseConnection

import java.sql.SQLException

class MealImpl extends MealInterface{


  override def findAllByUserId(id: Long): Seq[Meal] = {
    var conn: java.sql.Connection = null
    try {
      conn = DatabaseConnection.getConnection()
      val stmt = conn.prepareStatement("SELECT * FROM meal WHERE user_id = ?")
      stmt.setLong(1, id)
      val res = stmt.executeQuery()
      var seq: Seq[Meal] = null
      while (res.next()) {
        val v = new Meal(Option(res.getLong("id")), res.getInt("amount"), res.getLong("user_id"), res.getLong("food_id"), res.getInt("calories"), res.getInt("proteins"),
          res.getInt("fats"), res.getInt("carbohydrates"), res.getInt("fiber"), res.getBoolean("topicality"), res.getDate("date_created"))
        seq = seq :+ v
      }
     seq
    } catch {
      case e: SQLException =>
        println(s"SQL Exception: ${e.getMessage}")
        null
    }
    finally {
      if (conn != null) {
        try {
          conn.close()
        } catch {
          case e: SQLException =>
            e.printStackTrace()
        }
      }
    }
  }

  override def findAllTopical(id: Long): Seq[Meal] = {
    var conn: java.sql.Connection = null
    try {
      conn = DatabaseConnection.getConnection()
      val stmt = conn.prepareStatement("SELECT * FROM meal WHERE user_id = ? AND topicality = 0")
      stmt.setLong(1, id)
      val res = stmt.executeQuery()
      var seq: Seq[Meal] = Seq.empty[Meal]
      while (res.next()) {
        val v = new Meal(Option(res.getLong("id")), res.getDouble("amount"), res.getLong("user_id"), res.getLong("food_id"), res.getInt("calories"), res.getInt("proteins"),
          res.getInt("fats"), res.getInt("carbohydrates"), res.getInt("fiber"), res.getBoolean("topicality"), res.getDate("date_created"))
        seq = seq :+ v
      }
      seq
    } catch {
      case e: SQLException =>
        println(s"SQL Exception: ${e.getMessage}")
        null
    }
    finally {
      if (conn != null) {
        try {
          conn.close()
        } catch {
          case e: SQLException =>
            e.printStackTrace()
        }
      }
    }
  }

  override def deleteMeal(id: Long): Boolean = {
    var conn: java.sql.Connection = null
    try {
      conn = DatabaseConnection.getConnection()
      val stmt = conn.prepareStatement("DELETE FROM meal WHERE id = ?")
      stmt.setLong(1, id)
      stmt.executeUpdate()
      true
    } catch {
      case e: SQLException =>
        println(s"SQL Exception: ${e.getMessage}")
        false
    }
    finally {
      if (conn != null) {
        try {
          conn.close()
        } catch {
          case e: SQLException =>
            e.printStackTrace()
        }
      }
    }
  }

  override def addMeal(v: Meal): Boolean ={
    var conn: java.sql.Connection = null
    try {
      conn = DatabaseConnection.getConnection()
      val stmt = conn.prepareStatement("INSERT INTO meal (amount, user_id, food_id) VALUES (?, ?, ?)")
      stmt.setDouble(1, v.getAmount)
      stmt.setLong(2, v.getUserId)
      stmt.setLong(3, v.getFoodId)
      stmt.executeUpdate()
      true
    } catch {
      case e: SQLException =>
        println(s"SQL Exception: ${e.getMessage}")
        false
    }
    finally {
      if (conn != null) {
        try {
          conn.close()
        } catch {
          case e: SQLException =>
            e.printStackTrace()
        }
      }
    }
  }

  override def updateMeal(id: Long, amount: Double): Boolean = {
    var conn: java.sql.Connection = null
    try {
      conn = DatabaseConnection.getConnection()
      val stmt = conn.prepareStatement("UPDATE meal SET amount = ? WHERE id = ?")
      stmt.setDouble(1, amount)
      stmt.setLong(2, id)
      stmt.executeUpdate()
      true
    } catch {
      case e: SQLException =>
        println(s"SQL Exception: ${e.getMessage}")
        false
    }
    finally {
      if (conn != null) {
        try {
          conn.close()
        } catch {
          case e: SQLException =>
            e.printStackTrace()
        }
      }
    }
  }
}

File: lab3/src/main/java/dao/implementations/MealImpl.java
package dao.implementations;

import dao.interfaces.MealInterface;
import entities.Meal;

import jakarta.persistence.EntityManager;
import jakarta.persistence.EntityManagerFactory;
import jakarta.persistence.EntityTransaction;
import jakarta.persistence.TypedQuery;
import java.util.Collections;
import java.util.List;

public class MealImpl implements MealInterface {

    private final EntityManagerFactory emf;

    public MealImpl(EntityManagerFactory emf) {
        this.emf = emf;
    }

    @Override
    public List<Meal> findAllByUserId(Long id) {
        EntityManager em = emf.createEntityManager();
        try {
            TypedQuery<Meal> query = em.createQuery("SELECT m FROM Meal m WHERE m.userId = :userId", Meal.class);
            query.setParameter("userId", id);
            return query.getResultList();
        } catch(Exception e){
            e.printStackTrace();
            return Collections.emptyList();
        } finally {
            em.close();
        }
    }

    @Override
    public List<Meal> findAllTopical(Long id) {
        EntityManager em = emf.createEntityManager();
        try {
            TypedQuery<Meal> query = em.createQuery("SELECT m FROM Meal m WHERE m.userId = :userId AND m.topicality = false", Meal.class);
            query.setParameter("userId", id);
            return query.getResultList();
        } finally {
            em.close();
        }
    }

    @Override
    public boolean deleteMeal(Long id) {
        EntityManager em = emf.createEntityManager();
        EntityTransaction tx = em.getTransaction();
        try {
            tx.begin();
            Meal meal = em.find(Meal.class, id);
            if (meal != null) {
                em.remove(meal);
                tx.commit();
                return true;
            }
            return false;
        } catch (Exception e) {
            if (tx.isActive()) {
                tx.rollback();
            }
            e.printStackTrace();
            return false;
        } finally {
            em.close();
        }
    }

    @Override
    public boolean addMeal(Meal meal) {
        EntityManager em = emf.createEntityManager();
        EntityTransaction tx = em.getTransaction();
        try {
            tx.begin();
            em.persist(meal);
            tx.commit();
            return true;
        } catch (Exception e) {
            if (tx.isActive()) {
                tx.rollback();
            }
            e.printStackTrace();
            return false;
        } finally {
            em.close();
        }
    }

    @Override
    public boolean updateMeal(Long id, Double amount) {
        EntityManager em = emf.createEntityManager();
        EntityTransaction tx = em.getTransaction();
        try {
            tx.begin();
            Meal meal = em.find(Meal.class, id);
            if (meal != null) {
                meal.setAmount(amount);
                em.merge(meal);
                tx.commit();
                return true;
            }
            return false;
        } catch (Exception e) {
            if (tx.isActive()) {
                tx.rollback();
            }
            e.printStackTrace();
            return false;
        } finally {
            em.close();
        }
    }
}